civis/ml/_model.py

```diff
@@ -450,3 +462,518 @@ def training_metadata(self):
+                self._input_model = self.model  # Keep the estimator
+            self.model = str(estimator_file_id)
+        train_args['MODEL'] = self.model
+
+        name = self.model_name + ' Train' if self.model_name else None
+        # Clear the existing training result so we can make a new one.
+        self.train_result_ = None
+
+        result, container, run = self._create_custom_run(
+              self.train_template_id,
+              job_name=name,
+              table_name=table_name,
+              database_name=database_name,
+              file_id=file_id,
+              args=train_args,
+              resources=self.job_resources,
+              polling_interval=polling_interval)
+
+        self.train_result_ = result
+
+        return result
+
+    def _create_custom_run(self, template_id, job_name=None, table_name=None,
+                           database_name=None, file_id=None, args=None,
+                           resources=None, polling_interval=None):
+
+        db_id = self._client.get_database_id(database_name)
+        script_arguments = {'TABLE_NAME': table_name,
+                            'DB': {'database': db_id},
+                            'CIVIS_FILE_ID': file_id,
+                            'DEBUG': self.verbose}
+        resources = resources or {}
+        for key, value in resources.items():
+            if value:
+                # Default resources are set on the template. Only
+                # modify via arguments if users give a non-default value.
+                script_arguments[key] = value
+
+        script_arguments.update(args or {})
+
+        container = self._client.scripts.post_custom(
+            from_template_id=template_id,
+            name=job_name,
+            arguments=script_arguments)
+        log.info('Created custom script %s.', container.id)
+
+        run = self._client.scripts.post_custom_runs(container.id)
+        log.debug('Started job %s, run %s.', container.id, run.id)
+
+        train_kwargs = {}
+        if self.train_result_ is not None:
+            train_kwargs = {'train_job_id': self.train_result_.job_id,
+                            'train_run_id': self.train_result_.run_id}
+        fut = ModelFuture(
+              container.id,
+              run.id,
+              client=self._client,
+              polling_interval=polling_interval,
+              poll_on_creation=False,
+              **train_kwargs)
+
+        return fut, container, run
+
+    @property
+    @_check_is_fit
+    def state(self) -> str:
+        return self.train_result_.state
+
+    @property
+    @_check_is_fit
+    def estimator(self):
+        return self.train_result_.estimator
+
+    @_check_is_fit
+    def predict(self, X=None, table_name=None, database_name=None,
+                manifest=None, file_id=None, sql_where=None, sql_limit=None,
+                primary_key=SENTINEL, output_table=None, output_db=None,
```

Comment:
Is the use of `SENTINEL` here to allow users explicitly pass `None`?  And if so, what happens if `primary_key` is `None`?